markshust/docker-magento
# Improve bin/start and remove backslashes

### Description
1. Use `$(...)` (new style) to replace<code>`</code> backslashes (old style) because this [link](https://unix.stackexchange.com/questions/5778/whats-the-difference-between-stuff-and-stuff/5782#5782)

2. Improve `bin/start` to wait for all services of containers to be ready. 

No need the sleep any more `sleep 5 #Ensure containers are started...`

The concept is the same with #326 (Check Elasticsearch connection before setup:install)

I think the checker put to `bin/start` will be more fit, and also extend more services, such as nginx, mysql, redis, and rabbitmq

The checker also scans service ports to be available before docker start, and scan allocated ports by services after docker start.

If service ports are already allocated, `bin/start` will appear below error message.

<details>
  <summary>bin/start error for allocated port</summary>

```
bin/start
Starting a4f172b22c65_magento2test_rabbitmq_1 ...
Starting magento2test_elasticsearch_1         ...
Starting a4f172b22c65_magento2test_rabbitmq_1 ... error
Starting magento2test_redis_1                 ...
Starting magento2test_db_1                    ...
WARNING: Host is already in use by another container
WARNING: Host is already in use by another container

ERROR: for a4f172b22c65_magento2test_rabbitmq_1  Cannot start service rabbitmq: driver failed programming external connectivity on endpoint a4f172b22c65_magento2test_rabbitmq_1 (4a68cb572cc5641cea47c9292b1d0ce2edff8816f71f3a71e01e03884ad3a86c): Bind for 0.0.0.0:15672 failed: port is alStarting magento2test_db_1                    ... error

ERROR: for magento2test_db_1  Cannot start service db: driver failed programming external connectivity on endpoint magento2test_db_Starting magento2test_mailhog_1               ... error
WARNING: Host is already in use by another container
Starting magento2test_elasticsearch_1         ... error
ERROR: for magento2test_mailhog_1  Cannot start service mailhog: driver failed programming external connectivity on endpoint magento2test_mailhog_1 (71980bcf3e48e9b9fd8e21c433dfe0e70ae5418ff7bbc1eae54d386f555f0afb): Bind for 0.0.0.0:8025 failed: port is already allocated
WARNING: Host is already in use by another container

Starting magento2test_redis_1                 ... done
magento2test_elasticsearch_1 (81a677c15165102fcfdb505725ee695f8cb1eca3e3a52b5eacb88d5910eaacaa): Bind for 0.0.0.0:9300 failed: port is already allocated

ERROR: for rabbitmq  Cannot start service rabbitmq: driver failed programming external connectivity on endpoint a4f172b22c65_magento2test_rabbitmq_1 (4a68cb572cc5641cea47c9292b1d0ce2edff8816f71f3a71e01e03884ad3a86c): Bind for 0.0.0.0:15672 failed: port is already allocated

ERROR: for db  Cannot start service db: driver failed programming external connectivity on endpoint magento2test_db_1 (18268530f6de3196d087dfd681b1416916c23faa66414febdca298f437597837): Bind for 0.0.0.0:3306 failed: port is already allocated

ERROR: for mailhog  Cannot start service mailhog: driver failed programming external connectivity on endpoint magento2test_mailhog_1 (71980bcf3e48e9b9fd8e21c433dfe0e70ae5418ff7bbc1eae54d386f555f0afb): Bind for 0.0.0.0:8025 failed: port is already allocated

ERROR: for elasticsearch  Cannot start service elasticsearch: driver failed programming external connectivity on endpoint magento2test_elasticsearch_1 (81a677c15165102fcfdb505725ee695f8cb1eca3e3a52b5eacb88d5910eaacaa): Bind for 0.0.0.0:9300 failed: port is already allocated
ERROR: Encountered errors while bringing up the project.
```
</details>


### How to verify:
1. check onelinesetup and manual setup.
2. check `bin/start` and `bin/restart`
3. check `bin/xdebug enable` and `bin/xdebug disable`

Any concern please let me know, I will revise them.
